fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
-        process_group (optional): the c10d process group to be used for
-            distributed gradient reduction. If None, the default WORLD process group
-            will be used.
-        buffer_size (int, optional): number of elements to buffer before
-            performing reduce (default: 512k). Used to reduce multiple small
-            params to avoid communication overhead.
+    """
+    Wrap the model, and reduce the gradients to the right rank during the backward pass.
+
+    - the partition is given by the sharded optimizer
+    - wrap the base model with a model which knows where to reduce each gradient
+    - add an autograd function which calls the model grad dispatch on the way back
+
+     Args:
+        base_model (nn.Module):
+            model to be wrapped
+        sharded_optimizer (OSS, or list of OSS):
+            the sharded optimizer(s) which will decide the gradient partitioning
+
+    Keyword Args:
+        process_group (torch.nn.Optimizer):
+            optimizer to shard (default: SGD)
+        process_group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffers (bool):
+            whether to broadcast model buffers in between ranks at the beginning of each forward pass
+        reduce_buffer_size (int):
+            the size of the per-device-per-rank buffers used for the reduce operation
+
     """
 
     def __init__(
         self,
-        module: nn.Module,
-        optimizer: Type[torch.optim.Optimizer],
-        optimizer_params: Dict[str, Any],
-        world_size: int,
-        broadcast_buffers: bool,
+        base_model: nn.Module,
+        sharded_optimizer: Union[OSS, List[OSS]],
         process_group: Any = None,
-        buffer_size: int = 2 ** 19,
+        broadcast_buffers: bool = True,
     ):
         super().__init__()
 
-        self.module = module
-        self.world_size = world_size
+        self.base_model = base_model
+        self.sharded_optimizers = [sharded_optimizer] if isinstance(sharded_optimizer, OSS) else sharded_optimizer
+        self.enable_broadcast_buffers = broadcast_buffers
+        self.accumulate_grads = False
+
+        # Communication related attributes
         self.process_group = process_group if process_group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.process_group)
+        self.reference_global_rank = OSS.get_global_rank(self.process_group, 0)  # picking rank 0 as the reference
         self.rank = dist.get_rank(self.process_group)
-        self.broadcast_buffers = broadcast_buffers
-        self.authoritative_rank = 0
+        self.global_rank = OSS.get_global_rank(self.process_group, self.rank)
+
+        # Expose the same attributes as PytorchDDP, some frameworks rely on them.
+        # See https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel
+        self.devices = {p.device for p in self.base_model.parameters()}
```

Comment:
```suggestion
        self.device = {p.device for p in self.base_model.parameters()}
```
Looks like this variable is called `self.device` in [the link you posted](https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel).